beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationDryRunInvoice.java

```diff
@@ -0,0 +1,628 @@
+
+        // Specify bundleId filter
+        final DryRunArguments dryRunUpcomingInvoiceWithFilterArg3 = new TestDryRunArguments(DryRunType.UPCOMING_INVOICE, null, null, null, null, null, null, null, baseEntitlement.getBundleId(), null, null);
+        dryRunInvoice = invoiceUserApi.triggerInvoiceGeneration(account.getId(), null, dryRunUpcomingInvoiceWithFilterArg3, callContext);
+        assertEquals(dryRunInvoice.getTargetDate(), new LocalDate(2018, 2, 1));
+        invoiceChecker.checkInvoiceNoAudits(dryRunInvoice, callContext, expectedInvoices);
+
+    }
+
+
+
+    @Test(groups = "slow")
+    public void testDryRunWithUpcomingSubscriptionEvents() throws Exception {
+
+        final DateTime initialDate = new DateTime(2017, 11, 1, 0, 3, 42, 0, testTimeZone);
+
+        // set clock to the initial start date
+        clock.setDeltaFromReality(initialDate.getMillis() - clock.getUTCNow().getMillis());
+
+        final Account account = createAccountWithNonOsgiPaymentMethod(getAccountData(1));
+
+        final String productName = "Shotgun";
+        final BillingPeriod term = BillingPeriod.MONTHLY;
+
+        final DefaultEntitlement baseEntitlement = createBaseEntitlementAndCheckForCompletion(account.getId(), "bundleKey", productName, ProductCategory.BASE, term, NextEvent.CREATE, NextEvent.BLOCK, NextEvent.INVOICE);
+        assertNotNull(baseEntitlement);
+
+        // Verify the next invoice based on the PHASE event is correctly seen in the dryRun scenario
+        final List<ExpectedInvoiceItemCheck> expectedInvoices = new ArrayList<ExpectedInvoiceItemCheck>();
+        expectedInvoices.add(new ExpectedInvoiceItemCheck(new LocalDate(2017, 12, 1), new LocalDate(2018, 1, 1), InvoiceItemType.RECURRING, new BigDecimal("249.95")));
+        Invoice dryRunInvoice = invoiceUserApi.triggerInvoiceGeneration(account.getId(), null, DRY_RUN_UPCOMING_INVOICE_ARG, callContext);
+        invoiceChecker.checkInvoiceNoAudits(dryRunInvoice, callContext, expectedInvoices);
+        expectedInvoices.clear();
+
+
+        busHandler.pushExpectedEvents(NextEvent.PHASE, NextEvent.INVOICE, NextEvent.PAYMENT, NextEvent.INVOICE_PAYMENT);
+        clock.addDays(30);
+        assertListenerStatus();
+
+        invoiceChecker.checkInvoice(account.getId(), 2, callContext, new ExpectedInvoiceItemCheck(new LocalDate(2017, 12, 1), new LocalDate(2018, 1, 1), InvoiceItemType.RECURRING, new BigDecimal("249.95")));
+
+        // Future pause the subscription
+        LocalDate effectivePauseDate = new LocalDate(2017, 12, 15);
+        entitlementApi.pause(baseEntitlement.getBundleId(), effectivePauseDate, ImmutableList.<PluginProperty>of(), callContext);
+
+        // Verify the next invoice based on the PAUSE event is correctly seen in the dryRun scenario
+        expectedInvoices.add(new ExpectedInvoiceItemCheck(new LocalDate(2017, 12, 15), new LocalDate(2018, 1, 1), InvoiceItemType.REPAIR_ADJ, new BigDecimal("-137.07")));
```

Comment:
Would we see the same if we hadn't moved the clock? We probably wouldn't have a repair in that case?

Reply:
Don't understand the question -- in particular we have not yet moved the clock to reach the PAUSE date.

Reply:
The scenario is:

1. t=2017-11-1 CREATE
2. t=2017-12-1 PHASE (2017-12-1 -> 2018-1-1)
3. t=2017-12-1 PAUSE (future)
4. t=2017-12-15 PAUSE (2017-12-15 -> 2018-1-1 repair + CBA)

Now, if the scenario is:

1. t=2017-11-1 CREATE
2. t=2017-11-1 PAUSE (future)

and we trigger a dry-run: will we simply see a recurring item from 2017-12-1 to 2017-12-15 or do we see the expected invoice (recurring, repair, CBA)?